# New York Airbnb Analytics

### Data Processing For Group Project

# Setup

In [2]:

# Python ≥3.5 is required
import sys
assert sys.version_info >= (3, 5)

# Scikit-Learn ≥0.20 is required
import sklearn
assert sklearn.__version__ >= "0.20"

#Suppress all warning
import warnings
warnings.filterwarnings("ignore")
# Common imports
import numpy as np
import os

# To plot pretty figures
%matplotlib inline
import matplotlib as mpl
import matplotlib.pyplot as plt
mpl.rc('axes', labelsize=14)
mpl.rc('xtick', labelsize=12)
mpl.rc('ytick', labelsize=12)

# Ignore useless warnings (see SciPy issue #5998)

import warnings
warnings.filterwarnings(action="ignore", message="^internal gelsd")

#Display multiple outputs per cell

from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"


# Get the data

In [3]:
import pandas as pd
df1=pd.read_csv('http://data.insideairbnb.com/united-states/ny/new-york-city/2021-02-04/data/listings.csv.gz')
df1.head()

id                        listing_url       scrape_id last_scraped  \
0  2595  https://www.airbnb.com/rooms/2595  20210204180331   2021-02-05   
1  3831  https://www.airbnb.com/rooms/3831  20210204180331   2021-02-05   
2  5121  https://www.airbnb.com/rooms/5121  20210204180331   2021-02-05   
3  5178  https://www.airbnb.com/rooms/5178  20210204180331   2021-02-05   
4  5203  https://www.airbnb.com/rooms/5203  20210204180331   2021-02-06   

                                                name  \
0                              Skylit Midtown Castle   
1  Whole flr w/private bdrm, bath & kitchen(pls r...   
2                                    BlissArtsSpace!   
3                   Large Furnished Room Near B'way　   
4                 Cozy Clean Guest Room - Family Apt   

                                         description  \
0  Beautiful, spacious skylit studio in the heart...   
1  Enjoy 500 s.f. top floor in 1899 brownstone, w...   
2  <b>The space</b><br />HELLO EVERYONE AND THANK...   
3  Please don’t expect the luxury here just a bas...   
4  Our best guests are seeking a safe, clean, spa...   

                               neighborhood_overview  \
0  Centrally located in the heart of Manhattan ju...   
1  Just the right mix of urban center and local n...   
2                                                NaN   
3    Theater district, many restaurants around here.   
4  Our neighborhood is full of restaurants and ca...   

                                         picture_url  host_id  \
0  https://a0.muscache.com/pictures/f0813a11-40b2...     2845   
1  https://a0.muscache.com/pictures/e49999c2-9fd5...     4869   
2  https://a0.muscache.com/pictures/2090980c-b68e...     7356   
3  https://a0.muscache.com/pictures/12065/f070997...     8967   
4  https://a0.muscache.com/pictures/103776/b37157...     7490   

                                 host_url  ... review_scores_communication  \
0  https://www.airbnb.com/users/show/2845  ...                        10.0   
1  https://www.airbnb.com/users/show/4869  ...                        10.0   
2  https://www.airbnb.com/users/show/7356  ...                        10.0   
3  https://www.airbnb.com/users/show/8967  ...                         9.0   
4  https://www.airbnb.com/users/show/7490  ...                        10.0   

  review_scores_location review_scores_value license instant_bookable  \
0                   10.0                 9.0     NaN                f   
1                   10.0                10.0     NaN                f   
2                    9.0                 9.0     NaN                f   
3                   10.0                 9.0     NaN                f   
4                   10.0                10.0     NaN                f   

  calculated_host_listings_count calculated_host_listings_count_entire_homes  \
0                              2                                           2   
1                              1                                           1   
2                              1                                           0   
3                              1                                           0   
4                              1                                           0   

  calculated_host_listings_count_private_rooms  \
0                                            0   
1                                            0   
2                                            1   
3                                            1   
4                                            1   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.35  
1                                           0              4.99  
2                                           0              0.35  
3                                           0              3.31  
4                                           0              0.85  

[5 rows x 74 columns]

- Pick the valuable columns (not the columns with unique value for each row)

In [4]:
df1.drop(df1.columns[np.r_[0:3,6:10,12,18:22,24,26,34,31,42:55,56,61:68,69:73]], axis=1, inplace=True)
df1.head()

corr_matrix=df1.corr()
corr_matrix["review_scores_rating"].sort_values(ascending=False)

last_scraped                                               name  \
0   2021-02-05                              Skylit Midtown Castle   
1   2021-02-05  Whole flr w/private bdrm, bath & kitchen(pls r...   
2   2021-02-05                                    BlissArtsSpace!   
3   2021-02-05                   Large Furnished Room Near B'way　   
4   2021-02-06                 Cozy Clean Guest Room - Family Apt   

                                         description    host_name  host_since  \
0  Beautiful, spacious skylit studio in the heart...     Jennifer  2008-09-09   
1  Enjoy 500 s.f. top floor in 1899 brownstone, w...  LisaRoxanne  2008-12-07   
2  <b>The space</b><br />HELLO EVERYONE AND THANK...        Garon  2009-02-03   
3  Please don’t expect the luxury here just a bas...     Shunichi  2009-03-03   
4  Our best guests are seeking a safe, clean, spa...    MaryEllen  2009-02-05   

                                          host_about  host_response_time  \
0  A New Yorker since 2000! My passion is creatin...  within a few hours   
1  Laid-back Native New Yorker (formerly bi-coast...  within a few hours   
2   I am an artist(painter, filmmaker) and curato...                 NaN   
3  I used to work for a financial industry but no...        within a day   
4  Welcome to family life with my oldest two away...                 NaN   

  host_response_rate host_acceptance_rate host_is_superhost  ...    price  \
0                93%                  26%                 f  ...  $100.00   
1                98%                  93%                 f  ...   $73.00   
2                NaN                  NaN                 f  ...   $60.00   
3               100%                 100%                 f  ...   $79.00   
4                NaN                  NaN                 f  ...   $75.00   

  minimum_nights maximum_nights number_of_reviews number_of_reviews_l30d  \
0             30           1125                48                      0   
1              1            730               386                      3   
2             30            730                50                      0   
3              2             14               474                      0   
4              2             14               118                      0   

   first_review  last_review review_scores_rating  instant_bookable  \
0    2009-11-21   2019-11-04                 94.0                 f   
1    2014-09-30   2021-01-27                 90.0                 f   
2    2009-05-28   2019-12-02                 90.0                 f   
3    2009-05-06   2020-09-25                 84.0                 f   
4    2009-09-07   2017-07-21                 98.0                 f   

  reviews_per_month  
0              0.35  
1              4.99  
2              0.35  
3              3.31  
4              0.85  

[5 rows x 33 columns]

review_scores_rating         1.000000
number_of_reviews            0.054595
reviews_per_month            0.041548
bedrooms                     0.010334
number_of_reviews_l30d       0.005152
accommodates                -0.006905
beds                        -0.011301
maximum_nights              -0.013451
longitude                   -0.015705
minimum_nights              -0.025785
latitude                    -0.031618
host_total_listings_count   -0.032321
Name: review_scores_rating, dtype: float64

### 2. Clean/ Processing the dataset

In [5]:
#=======================================================Data Cleansing===================================================
#Only pick the Airbnb apartment with reviews
df1=df1[df1['number_of_reviews'].astype(int)>0]
#Only pick the Airbnb apartment with price
df1=df1[df1['price'].apply(lambda x: len(x)>0)]
#Only pick the Airbnb with the answer(t/f) for "host_is_superhost"
df1=df1[df1['host_is_superhost'].apply(lambda x: len(str(x))==1)]
#Drop other answer except f/t in "instant_bookable"
df1=df1[df1['instant_bookable'].isin(['f','t'])]
#To drop the review score which lower than 21 (potential outliers)
df1=df1[df1['review_scores_rating']>21]
#Replace all the blank cell with NaN value
df1=df1.replace('',np.NaN)
#========================================================Add new features===============================================
#Get the length of the sentence in following five columns (number of words)
df1['name_length']  = df1['name'].apply(lambda x: len(str(x).split()))
df1['description_length']=df1['description'].apply(lambda x: len(str(x).split()))
df1['host_about_length']=df1['host_about'].apply(lambda x: len(str(x).split()))
df1['verifications_length']=df1['host_verifications'].apply(lambda x: len(str(x).split(',')))
df1['amenities_length']=df1['amenities'].apply(lambda x: len(str(x).split(',')))
#Get the difference between the "last_scraped" with the following dates (in days)
df1['host_since_days'] = (pd.to_datetime(df1['last_scraped'])-pd.to_datetime(df1['host_since'])).dt.days
df1['first_reviews_days'] = (pd.to_datetime(df1['last_scraped'])-pd.to_datetime(df1['first_review'])).dt.days
df1['last_review_days'] = (pd.to_datetime(df1['last_scraped'])-pd.to_datetime(df1['last_review'])).dt.days

#Convert price column to the float datatype
df1['price'] = df1['price'].str.replace(',', '').str.replace('$', '').astype(float)
#Get the new column to express the price per accommodate
df1['price_per_accommodates']=df1['price']/df1['accommodates'].astype(float)
#=====================================================Change the data type=============================================
#Convert the categorical columns to dummified columns
list_col=['host_is_superhost','host_identity_verified','instant_bookable']
for i in list_col:
    df1[i] = np.where(df1[i]== 't', 1, 0)
    
#Add the dummified column by extract the word from "amenities" column
df1['has_Hair_dryer']=np.where(df1['amenities'].str.contains('Hair dryer'), 1, 0)
#df[df['date'].astype(str).str.contains('07311954')]

- Apply impute package to fill the missing data in numerical columns

In [6]:
print(df1.shape)

df1.info()

(26633, 43)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 26633 entries, 0 to 36915
Data columns (total 43 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   last_scraped                  26633 non-null  object 
 1   name                          26628 non-null  object 
 2   description                   26174 non-null  object 
 3   host_name                     26633 non-null  object 
 4   host_since                    26633 non-null  object 
 5   host_about                    17074 non-null  object 
 6   host_response_time            13574 non-null  object 
 7   host_response_rate            13574 non-null  object 
 8   host_acceptance_rate          17229 non-null  object 
 9   host_is_superhost             26633 non-null  int32  
 10  host_total_listings_count     26633 non-null  float64
 11  host_verifications            26633 non-null  object 
 12  host_identity_verified        26633 non-null  in

### Check for missing values

In [7]:
def missing_values(df):
    mis_val = df.isnull().sum()
    mis_val_percent = 100 * df.isnull().sum() / len(df)
    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
    mis_val_table_ren_columns = mis_val_table.rename(columns={
        0: 'Missing Values',
        1: '% of Total Values'
    })
    mis_val_table_ren_columns = mis_val_table_ren_columns[
        mis_val_table_ren_columns.iloc[:, 1] != 0].sort_values(
            '% of Total Values', ascending=False).round(1)
    print("Dataframe has " + str(df.shape[1]) + " columns.")
    print("There are " + str(mis_val_table_ren_columns.shape[0]) +
          " columns that have missing values.")
    
    return mis_val_table_ren_columns

# Missing values statistics
miss_values = missing_values(df1)
miss_values.head(20)

Dataframe has 43 columns.
There are 9 columns that have missing values.


Missing Values  % of Total Values
host_response_time             13059               49.0
host_response_rate             13059               49.0
host_about                      9559               35.9
host_acceptance_rate            9404               35.3
bedrooms                        2522                9.5
description                      459                1.7
beds                             130                0.5
bathrooms_text                    39                0.1
name                               5                0.0

### Flag Missing Values

**Create a new column called 'Flag' which is 1 if a row is complete with all its values, and 0 if there is a missing value in the row**

In [8]:
df1['flag'] = df1.notna().all(axis=1).view('i1')

df1

last_scraped                                               name  \
0       2021-02-05                              Skylit Midtown Castle   
1       2021-02-05  Whole flr w/private bdrm, bath & kitchen(pls r...   
2       2021-02-05                                    BlissArtsSpace!   
3       2021-02-05                   Large Furnished Room Near B'way　   
4       2021-02-06                 Cozy Clean Guest Room - Family Apt   
...            ...                                                ...   
36818   2021-02-06                       lanigirO    JFK•Casino•Beach   
36832   2021-02-05                Beautiful Room at Central Park West   
36848   2021-02-05            Big Blue Room & confortable in Brooklyn   
36851   2021-02-05                   Cozy room in heart of Brooklyn💓😌   
36915   2021-02-06  Peaceful & luxurious apt. in the heart of Astoria   

                                             description           host_name  \
0      Beautiful, spacious skylit studio in the heart...            Jennifer   
1      Enjoy 500 s.f. top floor in 1899 brownstone, w...         LisaRoxanne   
2      <b>The space</b><br />HELLO EVERYONE AND THANK...               Garon   
3      Please don’t expect the luxury here just a bas...            Shunichi   
4      Our best guests are seeking a safe, clean, spa...           MaryEllen   
...                                                  ...                 ...   
36818  This is an entire compact basement studio to y...                Ryan   
36832                                                NaN               Deniz   
36848  It is a beautiful private room, the bathroom i...  Cristina And Sasha   
36851                                                NaN           Valentina   
36915  Peaceful and Luxurious apartment in the heart ...            Giovanni   

       host_since                                         host_about  \
0      2008-09-09  A New Yorker since 2000! My passion is creatin...   
1      2008-12-07  Laid-back Native New Yorker (formerly bi-coast...   
2      2009-02-03   I am an artist(painter, filmmaker) and curato...   
3      2009-03-03  I used to work for a financial industry but no...   
4      2009-02-05  Welcome to family life with my oldest two away...   
...           ...                                                ...   
36818  2017-06-03                                                NaN   
36832  2015-08-03  I am Deniz 34 years old student pilot in New Y...   
36848  2016-11-15  Sasha is from Omaha originally but has lived i...   
36851  2019-10-07                                                NaN   
36915  2017-11-07                                                NaN   

       host_response_time host_response_rate host_acceptance_rate  \
0      within a few hours                93%                  26%   
1      within a few hours                98%                  93%   
2                     NaN                NaN                  NaN   
3            within a day               100%                 100%   
4                     NaN                NaN                  NaN   
...                   ...                ...                  ...   
36818      within an hour               100%                  98%   
36832      within an hour               100%                 100%   
36848      within an hour               100%                  98%   
36851  within a few hours                91%                 100%   
36915      within an hour               100%                 100%   

       host_is_superhost  ...  description_length host_about_length  \
0                      0  ...                 151                67   
1                      0  ...                 165                35   
2                      0  ...                 175                69   
3                      0  ...                 165                21   
4                      0  ...                 171               140   
...                  ...  ...                 ..

In [9]:
sum(df1.flag)

7831

There are 7831 rows with at least 1 missing value

**Another flagging method: flag the missing rows in each column separately**

In [10]:
cols = ['host_response_time','host_response_rate','host_acceptance_rate','bedrooms','description','beds','bathrooms_text','name']
df_try = df1[cols].isnull().astype(int).add_suffix('_indicator')
#df_try

#merge both the df1 and the flagged columns
df2 = pd.merge(df1, df_try, left_index=True, right_index=True)
df2.head(10)

last_scraped                                               name  \
0    2021-02-05                              Skylit Midtown Castle   
1    2021-02-05  Whole flr w/private bdrm, bath & kitchen(pls r...   
2    2021-02-05                                    BlissArtsSpace!   
3    2021-02-05                   Large Furnished Room Near B'way　   
4    2021-02-06                 Cozy Clean Guest Room - Family Apt   
5    2021-02-05     Lovely Room 1, Garden, Best Area, Legal rental   
6    2021-02-05                   Only 2 stops to Manhattan studio   
8    2021-02-06                            UES Beautiful Blue Room   
9    2021-02-05        Perfect for Your Parents:  Privacy + Garden   
11   2021-02-06                   Sweet and Spacious Brooklyn Loft   

                                          description      host_name  \
0   Beautiful, spacious skylit studio in the heart...       Jennifer   
1   Enjoy 500 s.f. top floor in 1899 brownstone, w...    LisaRoxanne   
2   <b>The space</b><br />HELLO EVERYONE AND THANK...          Garon   
3   Please don’t expect the luxury here just a bas...       Shunichi   
4   Our best guests are seeking a safe, clean, spa...      MaryEllen   
5   Discounted now!   Beautiful house, gorgeous ga...         Laurie   
6   Comfortable studio apartment with super comfor...  Allen & Irina   
8   Beautiful peaceful healthy home<br /><br /><b>...            Cyn   
9   Parents/grandparents coming to town, or are yo...           Jane   
11  A true open-plan loft in a repurposed factory ...          Chaya   

    host_since                                         host_about  \
0   2008-09-09  A New Yorker since 2000! My passion is creatin...   
1   2008-12-07  Laid-back Native New Yorker (formerly bi-coast...   
2   2009-02-03   I am an artist(painter, filmmaker) and curato...   
3   2009-03-03  I used to work for a financial industry but no...   
4   2009-02-05  Welcome to family life with my oldest two away...   
5   2009-03-10  Hello, \r\nI will be welcoming and helpful,  w...   
6   2009-05-06  We love to travel. When we travel we like to s...   
8   2009-05-12  Capturing the Steinbeck side of life in its Fi...   
9   2009-05-17  I have been an Airbnb host since 2009 -- just ...   
11  2009-06-12  We're a couple in our thirties who love to tra...   

    host_response_time host_response_rate host_acceptance_rate  \
0   within a few hours                93%                  26%   
1   within a few hours                98%                  93%   
2                  NaN                NaN                  NaN   
3         within a day               100%                 100%   
4                  NaN                NaN                  NaN   
5       within an hour               100%                 100%   
6                  NaN                NaN                  75%   
8       within an hour               100%                 100%   
9       within an hour               100%                 100%   
11      within an hour               100%                  62%   

    host_is_superhost  ...  has_Hair_dryer flag  host_response_time_indicator  \
0                   0  ...               1    0                             0   
1                   0  ...               1    1                             0   
2                   0  ...               0    0                             1   
3                   0  ...               1    1                             0   
4                   0  ...               1    0                             1   
5                   1  ...               1    1                             0   
6                   0  ...               1    0                             1   
8                   0  ...               1    1                             0   
9                   1  ...               1    1                             0   
11                  0  ...               0    0                             0   

   host_response_rate_indicator host_acceptance_rate_indicat

**Convert 'host_acceptance_rate' and 'host_response_rate' to numerical variables by removing the %'**

In [11]:
#get all the non-null values and Convert the object datatype to numerical datatype: host_response_rate, host_acceptance_rate
test = df2[['host_response_rate','host_acceptance_rate']].dropna()
test['host_acceptance_rate'] =test['host_acceptance_rate'].str.replace('%',"").astype(int)/100
test['host_response_rate'] =test['host_response_rate'].str.replace('%',"").astype(int)/100
test

#Add the numeric values back into the dataframe (replacement)
df2['host_response_rate'].loc[test.index] = test['host_response_rate']
df2['host_acceptance_rate'].loc[test.index] = test['host_acceptance_rate']
df2

host_response_rate  host_acceptance_rate
0                    0.93                  0.26
1                    0.98                  0.93
3                    1.00                  1.00
5                    1.00                  1.00
8                    1.00                  1.00
...                   ...                   ...
36818                1.00                  0.98
36832                1.00                  1.00
36848                1.00                  0.98
36851                0.91                  1.00
36915                1.00                  1.00

[13146 rows x 2 columns]

last_scraped                                               name  \
0       2021-02-05                              Skylit Midtown Castle   
1       2021-02-05  Whole flr w/private bdrm, bath & kitchen(pls r...   
2       2021-02-05                                    BlissArtsSpace!   
3       2021-02-05                   Large Furnished Room Near B'way　   
4       2021-02-06                 Cozy Clean Guest Room - Family Apt   
...            ...                                                ...   
36818   2021-02-06                       lanigirO    JFK•Casino•Beach   
36832   2021-02-05                Beautiful Room at Central Park West   
36848   2021-02-05            Big Blue Room & confortable in Brooklyn   
36851   2021-02-05                   Cozy room in heart of Brooklyn💓😌   
36915   2021-02-06  Peaceful & luxurious apt. in the heart of Astoria   

                                             description           host_name  \
0      Beautiful, spacious skylit studio in the heart...            Jennifer   
1      Enjoy 500 s.f. top floor in 1899 brownstone, w...         LisaRoxanne   
2      <b>The space</b><br />HELLO EVERYONE AND THANK...               Garon   
3      Please don’t expect the luxury here just a bas...            Shunichi   
4      Our best guests are seeking a safe, clean, spa...           MaryEllen   
...                                                  ...                 ...   
36818  This is an entire compact basement studio to y...                Ryan   
36832                                                NaN               Deniz   
36848  It is a beautiful private room, the bathroom i...  Cristina And Sasha   
36851                                                NaN           Valentina   
36915  Peaceful and Luxurious apartment in the heart ...            Giovanni   

       host_since                                         host_about  \
0      2008-09-09  A New Yorker since 2000! My passion is creatin...   
1      2008-12-07  Laid-back Native New Yorker (formerly bi-coast...   
2      2009-02-03   I am an artist(painter, filmmaker) and curato...   
3      2009-03-03  I used to work for a financial industry but no...   
4      2009-02-05  Welcome to family life with my oldest two away...   
...           ...                                                ...   
36818  2017-06-03                                                NaN   
36832  2015-08-03  I am Deniz 34 years old student pilot in New Y...   
36848  2016-11-15  Sasha is from Omaha originally but has lived i...   
36851  2019-10-07                                                NaN   
36915  2017-11-07                                                NaN   

       host_response_time host_response_rate host_acceptance_rate  \
0      within a few hours               0.93                 0.26   
1      within a few hours               0.98                 0.93   
2                     NaN                NaN                  NaN   
3            within a day                  1                    1   
4                     NaN                NaN                  NaN   
...                   ...                ...                  ...   
36818      within an hour                  1                 0.98   
36832      within an hour                  1                    1   
36848      within an hour                  1                 0.98   
36851  within a few hours               0.91                    1   
36915      within an hour                  1                    1   

       host_is_superhost  ...  has_Hair_dryer flag  \
0                      0  ...               1    0   
1                      0  ...               1    1   
2                      0  ...               0    0   
3                      0  ...               1    1   
4                      0  ...               1    0   
...                  ...  ...             ...  ...   
36818                  1  ...               0    0   
36832                  0  ...             

### Iterative imputer on numerical variables

In [12]:
#for some reason, the added values become 'objects', so you need to convert them to 'floats'
df2['host_response_rate'] = pd.to_numeric(df2['host_response_rate'],errors = 'coerce')
df2['host_acceptance_rate'] = pd.to_numeric(df2['host_acceptance_rate'],errors = 'coerce')

#choose numerical variables only
df_num=df2.drop(['neighbourhood_group_cleansed'],axis=1).iloc[:,np.r_[7:11,12,14:16,17,19:21,22:27,29:42]]

df_num

# Missing values statistics
miss_values = missing_values(df_num)
miss_values.head(20)

host_response_rate  host_acceptance_rate  host_is_superhost  \
0                    0.93                  0.26                  0   
1                    0.98                  0.93                  0   
2                     NaN                   NaN                  0   
3                    1.00                  1.00                  0   
4                     NaN                   NaN                  0   
...                   ...                   ...                ...   
36818                1.00                  0.98                  1   
36832                1.00                  1.00                  0   
36848                1.00                  0.98                  0   
36851                0.91                  1.00                  0   
36915                1.00                  1.00                  0   

       host_total_listings_count  host_identity_verified  latitude  longitude  \
0                            6.0                       1  40.75362  -73.98377   
1                            1.0                       1  40.68514  -73.95976   
2                            1.0                       1  40.68688  -73.95596   
3                            1.0                       0  40.76468  -73.98315   
4                            1.0                       1  40.80178  -73.96723   
...                          ...                     ...       ...        ...   
36818                        3.0                       1  40.67319  -73.84068   
36832                        4.0                       1  40.79851  -73.96004   
36848                        4.0                       1  40.70146  -73.94368   
36851                        1.0                       1  40.70430  -73.92561   
36915                        1.0                       1  40.76801  -73.91197   

       accommodates  bedrooms  beds  ...  name_length  description_length  \
0                 2       NaN   1.0  ...            3                 151   
1                 3       1.0   3.0  ...            8                 165   
2                 2       1.0   1.0  ...            1                 175   
3                 2       1.0   1.0  ...            5                 165   
4                 1       1.0   1.0  ...            7                 171   
...             ...       ...   ...  ...          ...                 ...   
36818             2       1.0   1.0  ...            2                 168   
36832             1       4.0   5.0  ...            6                   1   
36848             2       1.0   1.0  ...            7                 146   
36851             2       1.0   1.0  ...            6                   1   
36915             3       1.0   1.0  ...            9                  72   

       host_about_length  verifications_length  amenities_length  \
0                     67                     9                31   
1                     35                     6                35   
2                     69                     8                 5   
3                     21                     4                18   
4                    140                     7                16   
...                  ...                   ...               ...   
36818                  1                    10                16   
36832                 23                     5                20   
36848                221                     7                28   
36851                  1                     6                20   
36915                  1                     4                19   

       host_since_days  first_reviews_days  last_review_days  \
0                 4532                4094               459   
1                 4443                2320                 9   
2                 4385                4271               431   
3                 4357                4293               133   
4                 4384                4170              1296   
...                ...                 ...               ...   
368

Dataframe has 28 columns.
There are 4 columns that have missing values.


Missing Values  % of Total Values
host_response_rate             13487               50.6
host_acceptance_rate           13487               50.6
bedrooms                        2522                9.5
beds                             130                0.5

In [13]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

imp = IterativeImputer(random_state=0)

df_num1 = imp.fit_transform(df_num)

cols = list(df_num)

df_num1=pd.DataFrame(df_num1)
df_num1.columns=cols
df_num1

# Re-check Missing values statistics
miss_values = missing_values(df_num1)
miss_values.head(20)

host_response_rate  host_acceptance_rate  host_is_superhost  \
0                0.930000              0.260000                0.0   
1                0.980000              0.930000                0.0   
2                0.922288              0.575276                0.0   
3                1.000000              1.000000                0.0   
4                0.880550              0.498145                0.0   
...                   ...                   ...                ...   
26628            1.000000              0.980000                1.0   
26629            1.000000              1.000000                0.0   
26630            1.000000              0.980000                0.0   
26631            0.910000              1.000000                0.0   
26632            1.000000              1.000000                0.0   

       host_total_listings_count  host_identity_verified  latitude  longitude  \
0                            6.0                     1.0  40.75362  -73.98377   
1                            1.0                     1.0  40.68514  -73.95976   
2                            1.0                     1.0  40.68688  -73.95596   
3                            1.0                     0.0  40.76468  -73.98315   
4                            1.0                     1.0  40.80178  -73.96723   
...                          ...                     ...       ...        ...   
26628                        3.0                     1.0  40.67319  -73.84068   
26629                        4.0                     1.0  40.79851  -73.96004   
26630                        4.0                     1.0  40.70146  -73.94368   
26631                        1.0                     1.0  40.70430  -73.92561   
26632                        1.0                     1.0  40.76801  -73.91197   

       accommodates  bedrooms  beds  ...  name_length  description_length  \
0               2.0  1.048801   1.0  ...          3.0               151.0   
1               3.0  1.000000   3.0  ...          8.0               165.0   
2               2.0  1.000000   1.0  ...          1.0               175.0   
3               2.0  1.000000   1.0  ...          5.0               165.0   
4               1.0  1.000000   1.0  ...          7.0               171.0   
...             ...       ...   ...  ...          ...                 ...   
26628           2.0  1.000000   1.0  ...          2.0               168.0   
26629           1.0  4.000000   5.0  ...          6.0                 1.0   
26630           2.0  1.000000   1.0  ...          7.0               146.0   
26631           2.0  1.000000   1.0  ...          6.0                 1.0   
26632           3.0  1.000000   1.0  ...          9.0                72.0   

       host_about_length  verifications_length  amenities_length  \
0                   67.0                   9.0              31.0   
1                   35.0                   6.0              35.0   
2                   69.0                   8.0               5.0   
3                   21.0                   4.0              18.0   
4                  140.0                   7.0              16.0   
...                  ...                   ...               ...   
26628                1.0                  10.0              16.0   
26629               23.0                   5.0              20.0   
26630              221.0                   7.0              28.0   
26631                1.0                   6.0              20.0   
26632                1.0                   4.0              19.0   

       host_since_days  first_reviews_days  last_review_days  \
0               4532.0              4094.0             459.0   
1               4443.0              2320.0               9.0   
2               4385.0              4271.0             431.0   
3               4357.0              4293.0             133.0   
4               4384.0              4170.0            1296.0   
...                ...                 ...               ...   
266

Dataframe has 28 columns.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []

**Now, replace the incomplete columns in df2 with the corresponding imputed columns from df_num1**

In [14]:
df2[cols] = df_num1[cols].values

df2.head(5)

last_scraped                                               name  \
0   2021-02-05                              Skylit Midtown Castle   
1   2021-02-05  Whole flr w/private bdrm, bath & kitchen(pls r...   
2   2021-02-05                                    BlissArtsSpace!   
3   2021-02-05                   Large Furnished Room Near B'way　   
4   2021-02-06                 Cozy Clean Guest Room - Family Apt   

                                         description    host_name  host_since  \
0  Beautiful, spacious skylit studio in the heart...     Jennifer  2008-09-09   
1  Enjoy 500 s.f. top floor in 1899 brownstone, w...  LisaRoxanne  2008-12-07   
2  <b>The space</b><br />HELLO EVERYONE AND THANK...        Garon  2009-02-03   
3  Please don’t expect the luxury here just a bas...     Shunichi  2009-03-03   
4  Our best guests are seeking a safe, clean, spa...    MaryEllen  2009-02-05   

                                          host_about  host_response_time  \
0  A New Yorker since 2000! My passion is creatin...  within a few hours   
1  Laid-back Native New Yorker (formerly bi-coast...  within a few hours   
2   I am an artist(painter, filmmaker) and curato...                 NaN   
3  I used to work for a financial industry but no...        within a day   
4  Welcome to family life with my oldest two away...                 NaN   

   host_response_rate  host_acceptance_rate  host_is_superhost  ...  \
0            0.930000              0.260000                0.0  ...   
1            0.980000              0.930000                0.0  ...   
2            0.922288              0.575276                0.0  ...   
3            1.000000              1.000000                0.0  ...   
4            0.880550              0.498145                0.0  ...   

   has_Hair_dryer flag  host_response_time_indicator  \
0             1.0    0                             0   
1             1.0    1                             0   
2             0.0    0                             1   
3             1.0    1                             0   
4             1.0    0                             1   

  host_response_rate_indicator host_acceptance_rate_indicator  \
0                            0                              0   
1                            0                              0   
2                            1                              1   
3                            0                              0   
4                            1                              1   

   bedrooms_indicator  description_indicator beds_indicator  \
0                   1                      0              0   
1                   0                      0              0   
2                   0                      0              0   
3                   0                      0              0   
4                   0                      0              0   

   bathrooms_text_indicator name_indicator  
0                         0              0  
1                         0              0  
2                         1              0  
3                         0              0  
4                         0              0  

[5 rows x 52 columns]

In [15]:
# Check Missing values statistics again
miss_values = missing_values(df2)
miss_values.head(20)

Dataframe has 52 columns.
There are 5 columns that have missing values.


Missing Values  % of Total Values
host_response_time           13059               49.0
host_about                    9559               35.9
description                    459                1.7
bathrooms_text                  39                0.1
name                             5                0.0

### We have taken care of the numerical values, and need to work on the categorical values next...

In [16]:
df2['host_response_time'].mode()
df2.shape

0    within an hour
dtype: object

(26633, 52)

In [17]:
df2['host_about'] = df2['host_about'].fillna("Unknown")
df2['host_response_time'] = df2['host_response_time'].fillna("within an hour")

In [18]:

df2 = df2.dropna(axis=0) 

df2.shape

# Check Missing values statistics again
miss_values = missing_values(df2)
miss_values.head(20)

(26137, 52)

Dataframe has 52 columns.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []

### 3. Train test split

In [19]:
df2.select_dtypes(include=[object]).columns

Index(['last_scraped', 'name', 'description', 'host_name', 'host_since',
       'host_about', 'host_response_time', 'host_verifications',
       'neighbourhood_cleansed', 'neighbourhood_group_cleansed', 'room_type',
       'bathrooms_text', 'amenities', 'first_review', 'last_review'],
      dtype='object')

In [22]:
df_split = df2[['host_response_rate',
       'host_acceptance_rate', 'host_is_superhost',
       'host_total_listings_count', 
       'host_identity_verified', 'latitude', 'longitude',
       'accommodates', 'bedrooms', 'beds', 
       'price', 'minimum_nights', 'maximum_nights', 'number_of_reviews',
       'number_of_reviews_l30d',
       'review_scores_rating', 'instant_bookable', 'reviews_per_month',
       'name_length', 'description_length', 'host_about_length',
       'verifications_length', 'amenities_length', 'host_since_days',
       'first_reviews_days', 'last_review_days', 'price_per_accommodates',
       'has_Hair_dryer', 'flag', 'host_response_time_indicator',
       'host_response_rate_indicator', 'host_acceptance_rate_indicator',
       'bedrooms_indicator', 'description_indicator', 'beds_indicator',
       'bathrooms_text_indicator', 'name_indicator']]

In [23]:
from sklearn.model_selection import train_test_split

train_set, test_set = train_test_split(df_split, test_size=0.2, random_state=42)

test_set.head()

host_response_rate  host_acceptance_rate  host_is_superhost  \
23984            0.670000              1.000000                0.0   
1707             0.500000              0.830000                0.0   
14244            0.818969              0.687823                0.0   
22627            1.000000              0.890000                1.0   
2136             1.000000              0.970000                0.0   

       host_total_listings_count  host_identity_verified  latitude  longitude  \
23984                       11.0                     1.0  40.75643  -73.98647   
1707                         8.0                     1.0  40.75639  -73.96558   
14244                        1.0                     1.0  40.76152  -73.99897   
22627                        1.0                     0.0  40.82560  -73.95187   
2136                         1.0                     1.0  40.69594  -73.96715   

       accommodates  bedrooms  beds  ...  has_Hair_dryer  flag  \
23984           2.0       1.0   1.0  ...             1.0     1   
1707            4.0       1.0   2.0  ...             1.0     1   
14244           3.0       1.0   2.0  ...             1.0     0   
22627           1.0       1.0   1.0  ...             1.0     1   
2136            2.0       1.0   1.0  ...             1.0     1   

       host_response_time_indicator  host_response_rate_indicator  \
23984                             0                             0   
1707                              0                             0   
14244                             1                             1   
22627                             0                             0   
2136                              0                             0   

       host_acceptance_rate_indicator  bedrooms_indicator  \
23984                               0                   0   
1707                                0                   0   
14244                               1                   0   
22627                               0                   0   
2136                                0                   0   

       description_indicator  beds_indicator  bathrooms_text_indicator  \
23984                      0               0                         0   
1707                       0               0                         0   
14244                      0               0                         0   
22627                      0               0                         0   
2136                       0               0                         0   

       name_indicator  
23984               0  
1707                0  
14244               0  
22627               0  
2136                0  

[5 rows x 37 columns]

## ML FLOW

### Attempt 1

In [29]:
from math import exp
import xgboost as xgb
!pip install hyperopt
from hyperopt import fmin, hp, tpe, SparkTrials, STATUS_OK
import mlflow
from sklearn.preprocessing import StandardScaler

In [31]:
import numpy as np
import pandas as pd
import xgboost as xg
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error as MSE
  
# Load the data
X_train = train_set.drop(columns=['price'], axis=1)
y_train = train_set['price']

X_test = test_set.drop(columns=['price'], axis=1)
y_test = test_set['price']

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)
  
# Instantiation
xgb_r = xg.XGBRegressor(objective ='reg:squarederror',
                  n_estimators = 10, seed = 123)
  
# Fitting the model
xgb_r.fit(X_train, y_train)
  
# Predict the model
pred = xgb_r.predict(X_test)
  
# RMSE Computation
rmse = np.sqrt(MSE(y_test, pred))
print("RMSE : % f" %(rmse))\

#ML Flow
with mlflow.start_run():
  mlflow.log_param('alpha','0.1')
  mlflow.log_metric('mae', 0.05)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints='',
             learning_rate=0.300000012, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=10, n_jobs=8, num_parallel_tree=1, random_state=123,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=123,
             subsample=1, tree_method='exact', validate_parameters=1,
             verbosity=None)

RMSE :  85.535391


### Attempt 2

In [33]:
def params_to_xgb(params):
  return{
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': int(params['max_depth']),
    'learning_rate': exp(params['log_learning_rate']),
    'reg_alpha': exp(params['log_reg_alpha']),
    'reg_lambda': exp(params['log_reg_lambda']),
    'gamma': exp(params['log_gamma']),
    'min_child_weight': exp(params['log_min_child_weight']),
   # 'importance_type': 'total_gain',
    'seed': 0
  }


X_train = train_set.drop(columns=['price'], axis=1)
y_train = train_set['price']

X_test = test_set.drop(columns=['price'], axis=1)
y_test = test_set['price']

sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.fit_transform(X_test)

def train_model(params):
  train=xgb.DMatrix(data=X_train, label=y_train)
  test=xgb.DMatrix(data=X_test, label=y_test)
  booster = xgb.train(params=params_to_xgb(params), dtrain=X_train, num_boost_round=1000,
                     evals=[(X_test, "rmse")], early_stopping_rounds=50)
  mlflow.log_param('best_iteration', booster.attr('best_iteration'))
  return {'status':STATUS_OK, 'loss': booster.best_score, 'booster':booster.attributes()}

search_space = {
  'max_depth': hp.quniform('max_depth', 20, 60, 1),
  'learning_rate': hp.uniform('log_learning_rate', -3, 0),
  'reg_alpha': hp.uniform('log_reg_alpha', -5, -1),
  'reg_lambda': hp.uniform('log_reg_lambda', 1, 8),
  'gamma': hp.uniform('log_gamma', -6, -1),
  'min_child_weight': hp.uniform('log_min_child_weight', -1, 3),
}

spark_trials = SparkTrials(parallelism=12)
best_params = fmin(fn=train_model, space=search_space, algo=tpe.suggest, max_evals=96, trials=spark_trials, rstate=np.random.RandomState(123))

Exception: Java gateway process exited before sending its port number

### Attempt 3

In [28]:
#!pip install mlflow
import os
import warnings
import sys

import pandas as pd
import numpy as np
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.linear_model import ElasticNet
from urllib.parse import urlparse
import mlflow
import mlflow.sklearn

import logging

logging.basicConfig(level=logging.WARN)
logger = logging.getLogger(__name__)


def eval_metrics(actual, pred):
    rmse = np.sqrt(mean_squared_error(actual, pred))
    mae = mean_absolute_error(actual, pred)
    r2 = r2_score(actual, pred)
    return rmse, mae, r2


if __name__ == "__main__":
    warnings.filterwarnings("ignore")
    np.random.seed(40)

    
    
    train_x = train_set.drop(columns=['price'], axis=1)
    train_y = train_set['price']

    test_x = test_set.drop(columns=['price'], axis=1)
    test_y = test_set['price']


    alpha = float(sys.argv[1]) if len(sys.argv) > 1 else 0.5
    l1_ratio = float(sys.argv[2]) if len(sys.argv) > 2 else 0.5

    with mlflow.start_run():
        lr = ElasticNet(alpha=alpha, l1_ratio=l1_ratio, random_state=42)
        lr.fit(train_x, train_y)

        predicted_qualities = lr.predict(test_x)

        (rmse, mae, r2) = eval_metrics(test_y, predicted_qualities)

        print("Elasticnet model (alpha=%f, l1_ratio=%f):" % (alpha, l1_ratio))
        print("  RMSE: %s" % rmse)
        print("  MAE: %s" % mae)
        print("  R2: %s" % r2)

        mlflow.log_param("alpha", alpha)
        mlflow.log_param("l1_ratio", l1_ratio)
        mlflow.log_metric("rmse", rmse)
        mlflow.log_metric("r2", r2)
        mlflow.log_metric("mae", mae)

        tracking_url_type_store = urlparse(mlflow.get_tracking_uri()).scheme

        # Model registry does not work with file store
        if tracking_url_type_store != "file":

            # Register the model
            # There are other ways to use the Model Registry, which depends on the use case,
            # please refer to the doc for more information:
            # https://mlflow.org/docs/latest/model-registry.html#api-workflow
            mlflow.sklearn.log_model(lr, "model", registered_model_name="ElasticnetWineModel")
        else:
            mlflow.sklearn.log_model(lr, "model")

ValueError: could not convert string to float: '-f'

In [ ]:
edit edit edit